ray-on-gke/tpu/kuberay-tpu-webhook/webhook_main_test.go

```diff
@@ -301,25 +306,80 @@ func setupTest(t *testing.T) {
 			},
 		},
 	}
+
+	// add 4 TPU worker pods with env vars set
+	numOfHosts := 2
+	numSlices := 2
+	podIndex := 0
+	testInterceptedTPUPods = []*corev1.Pod{}
+	for i := 0; i < numSlices; i++ {
+		// generate hostnames for this slice
+		hostnames := make([]string, numOfHosts)
+		for ind := 0; ind < numOfHosts; ind++ {
+			hostnames[i] = fmt.Sprintf("%s-%d-%d", groupNameStr, i, ind)
+		}
+		testHostnames := strings.Join(hostnames, ",")
+		replicaIndex := fmt.Sprintf("%s-%d", groupNameStr, i)
+		for j := 0; j < numOfHosts; j++ {
+			testTPUWorkerCopy := testTPUWorker.DeepCopy()
+			// set TPU environment variables for this Pod
+			env := []corev1.EnvVar{
+				{
+					Name:  "TPU_WORKER_ID",
+					Value: fmt.Sprint(j),
+				},
+				{
+					Name:  "TPU_WORKER_HOSTNAMES",
+					Value: testHostnames,
+				},
+				{
+					Name:  "TPU_NAME",
+					Value: replicaIndex,
+				},
+			}
+			testTPUWorkerCopy.Spec.Containers[0].Env = env
+			testTPUWorkerCopy.Name = fmt.Sprintf("%s-%d", "intercepted-tpu-pod", podIndex)
+			podIndex += 1
+			testTPUWorkerCopy.Labels["replicaIndex"] = replicaIndex
+			testInterceptedTPUPods = append(testInterceptedTPUPods, testTPUWorkerCopy)
+		}
+	}
 }
 
-// helper function used by tests which mutate sliceToHostnames
-func deepCopySliceToHostnames() map[slice]string {
-	deepCopy := make(map[slice]string)
-	for slice, hostnames := range sliceToHostnames {
-		deepCopy[slice] = hostnames
+func setupInformer(pods []*corev1.Pod) {
+	// initialize fake Clientset with pod objects
+	tpuObjects := make([]runtime.Object, len(pods))
+	for i, pod := range pods {
+		tpuObjects[i] = pod
 	}
+	fakeClientSet := fake.NewSimpleClientset(tpuObjects...)
 
-	return deepCopy
+	// initialize podLister using the fake client for testing
+	factory := informers.NewSharedInformerFactory(fakeClientSet, 0)
+	podInformer := factory.Core().V1().Pods().Informer()
+
+	stopCh := make(chan struct{})
+	defer close(stopCh)
+
+	factory.Start(stopCh)
+	factory.WaitForCacheSync(stopCh)
+
+	// wait for cache to sync before creating the Lister
+	if !cache.WaitForCacheSync(stopCh, podInformer.HasSynced) {
+		fmt.Printf("Timed out waiting for fake client to sync")
+		return
+	}
+
+	podLister = factory.Core().V1().Pods().Lister()
```

Comment:
Each test is trying to override and use podLIster, which is a global variable in `main.go`. It will lead to flaky or unexpected behavior in the tests, especially if they run in paralllel.

Reply:
Each test should be initializing it's own informer / pod lister so that only the pod objects relelvant for the test are in the cache

Reply:
Done in [8efed67](https://github.com/GoogleCloudPlatform/ai-on-gke/pull/723/commits/8efed67ecf27a6368d41648f32da6e44b732cc6c), I refactored the webhook to not use global vars and to pass the `PodLister` object as part of a new `TPUWebhookServer`.